src/codewhisperer/util/closingBracketUtil.ts

```diff
@@ -4,71 +4,123 @@
  */
 
 import * as vscode from 'vscode'
-import { TextEdit, workspace, WorkspaceEdit } from 'vscode'
+import { workspace, WorkspaceEdit } from 'vscode'
 import { isCloud9 } from '../../shared/extensionUtilities'
-import * as telemetry from '../../shared/telemetry/telemetry'
+import { CodeWhispererConstants } from '../models/constants'
 
-export async function handleAutoClosingBrackets(
-    triggerType: telemetry.CodewhispererTriggerType,
-    editor: vscode.TextEditor,
-    recommendation: string,
-    line: number,
-    specialChar: string
-) {
-    const openingBrackets = new Map<string, string>()
-    openingBrackets.set('{', '}')
-    openingBrackets.set('(', ')')
-    openingBrackets.set('[', ']')
-    const closingBracket = openingBrackets.get(specialChar)
-    if (
-        triggerType === 'AutoTrigger' &&
-        closingBracket !== undefined &&
-        hasExtraClosingBracket(recommendation, specialChar, closingBracket)
-    ) {
-        let curLine = line
-        let textLine = editor.document.lineAt(curLine).text
-        while (curLine > 0 && !textLine.includes(closingBracket)) {
-            curLine--
-            textLine = editor.document.lineAt(curLine).text
+interface bracketMapType {
+    [k: string]: string
+}
+
+const bracketMap: bracketMapType = {
+    ')': '(',
+    ']': '[',
+    '}': '{',
+}
+
+export const calculateBracketsLevel = (code: string) => {
+    const bracketCounts: Map<string, number> = new Map([
+        ['(', 0],
+        ['[', 0],
+        ['{', 0],
+    ])
+    const bracketsIndexLevel = []
+
+    for (let i = 0; i < code.length; i++) {
```

Comment:
This will go through at most 10k characters.

Reply:
I tested with a 1000+ line right context and didn't see any performance issue, but I can probably lower this number

Reply:
An optimization of this could be taking another string into account. Here we are calculating the bracket levels of the recommendation and the right context independently; One way to achieve this is to match the brackets on the fly when we iterate through recommendations, so the algorithm could be:
```
iterate through each char of recommendation:
    c = the current character in right context that needs to be bracket or space-matched
    if (c is not bracket and is not empty character):
        break
    // we want to match c in recommendation
    current = current char of recommendation
    if (current matches c both character-wise and level-wise):
        c++
        record this index to be processed further
```

Reply:
This is non-blocking for this PR, we can get this merged first and have a sim to work on the optimization later.

Reply:
updated to return early when processing right context and we encounter a non-bracket, non-white-space character